File: src/transaction.zig
const std = @import("std");
const hex = std.fmt.fmtSliceHexLower;
const expectEqualSlices = std.testing.expectEqualSlices;
const assert = std.debug.assert;
const Blake3 = std.crypto.hash.Blake3;

const lmdb = @import("lmdb");
const utils = @import("utils.zig");
const fmtKey = utils.fmtKey;

const Result = enum { update, delete };

const nil = [0]u8{};

pub fn Transaction(comptime K: u8, comptime Q: u32) type {
    const Node = @import("node.zig").Node(K, Q);
    const Tree = @import("tree.zig").Tree(K, Q);
    const Cursor = @import("cursor.zig").Cursor(K, Q);
    const Header = @import("header.zig").Header(K, Q);
    const Logger = @import("logger.zig").Logger;
    const NodeList = @import("node_list.zig").NodeList(K, Q);
    const BufferPool = @import("buffer_pool.zig").BufferPool;
    const NodeEncoder = @import("node_encoder.zig").NodeEncoder(K, Q);

    return struct {
        const Self = @This();
        pub const Mode = enum { ReadOnly, ReadWrite };
        pub const Effects = struct { create: usize = 0, update: usize = 0, delete: usize = 0, height: u8 = 0 };
        pub const Options = struct {
            mode: Mode,
            dbi: ?[*:0]const u8 = null,
            log: ?std.fs.File.Writer = null,
            trace: ?*NodeList = null,
            effects: ?*Effects = null,
        };

        const OperationType = enum { set, delete };
        const OperationTag = union(OperationType) {
            set: struct { key: []const u8, value: []const u8 },
            delete: []const u8,
        };

        const Operation = struct {
            pub fn Set(key: []const u8, value: []const u8) OperationTag {
                return .{ .set = .{ .key = key, .value = value } };
            }

            pub fn Delete(key: []const u8) OperationTag {
                return .{ .delete = key };
            }

            tag: OperationTag,
            ctx: *Self,

            txn: lmdb.Transaction,
            cursor: Cursor,
            logger: Logger,

            trace: ?*NodeList = null,
            effects: ?*Effects = null,

            encoder: *NodeEncoder,
            pool: *BufferPool,
            new_siblings: *std.ArrayList(?[]const u8),
            hash_buffer: [K]u8 = undefined,

            pub fn init(allocator: std.mem.Allocator, ctx: *Self, tag: OperationTag) !Operation {
                const txn = try lmdb.Transaction.open(ctx.env, .{
                    .read_only = false,
                    .parent = ctx.txn,
                    .dbi = ctx.dbi,
                });

                const cursor = try Cursor.open(allocator, txn);
                return Operation{
                    .tag = tag,
                    .ctx = ctx,
                    .txn = txn,
                    .cursor = cursor,
                    .trace = ctx.trace,
                    .effects = ctx.effects,
                    .logger = Logger.init(allocator, ctx.log),

                    .encoder = &ctx.encoder,
                    .pool = &ctx.pool,
                    .new_siblings = &ctx.new_siblings,

                    // .encoder = NodeEncoder.init(allocator),
                    // .pool = BufferPool.init(allocator),
                    // .new_siblings = std.ArrayList(?[]const u8).init(allocator),
                };
            }

            pub fn abort(self: *Operation) void {
                self.deinit();
                self.cursor.close();
                self.txn.abort();
            }

            pub fn commit(self: *Operation) !void {
                self.deinit();
                self.cursor.close();
                try self.txn.commit();
            }

            pub fn deinit(self: *Operation) void {
                self.logger.deinit();
                // self.new_siblings.deinit();
                // self.encoder.deinit();
                // self.pool.deinit();
            }

            pub fn apply(self: *Operation) !void {
                if (self.trace) |trace| trace.reset();
                if (self.effects) |effects| effects.* = Effects{};

                const root = try self.cursor.goToRoot();
                try self.log("height: {d}", .{root.level});

                var root_level = if (root.level == 0) 1 else root.level;

                self.logger.reset();
                try self.new_siblings.resize(0);
                try self.pool.allocate(root_level - 1);

                const result = try switch (root.level) {
                    0 => self.applyLeaf(null),
                    else => self.applyNode(root_level - 1, null),
                };

                try switch (result) {
                    Result.update => {},
                    Result.delete => error.InternalError,
                };

                _ = try self.hashNode(root_level, null);

                try self.log("new_children: {d}", .{self.new_siblings.items.len});
                for (self.new_siblings.items) |child| {
                    try self.log("- {s}", .{fmtKey(child)});
                }

                while (self.new_siblings.items.len > 0) {
                    try self.promote(root_level);

                    root_level += 1;
                    _ = try self.hashNode(root_level, null);
                    try self.log("new_children: {d}", .{self.new_siblings.items.len});
                    for (self.new_siblings.items) |child| {
                        try self.log("- {s}", .{fmtKey(child)});
                    }
                }

                while (root_level > 0) : (root_level -= 1) {
                    _ = try self.cursor.goToNode(root_level - 1, null);
                    if (try self.cursor.goToNext()) |_| {
                        break;
                    } else {
                        try self.log("trim root from {d} to {d}", .{ root_level, root_level - 1 });
                        try self.deleteNode(root_level, null);
                    }
                }

                if (self.effects) |effects| effects.height = root_level + 1;
            }

            fn applyLeaf(self: *Operation, first_child: ?[]const u8) !Result {
                switch (self.tag) {
                    .set => |tag| {
                        utils.hashEntry(tag.key, tag.value, &self.hash_buffer);
                        const leaf = Node{
                            .level = 0,
                            .key = tag.key,
                            .hash = &self.hash_buffer,

                            // TODO: wtf?
                            // some weird bug causes .value to be null if entry.value.len == 0.
                            // setting it to another explicit empty slice seems to fix it.
                            .value = if (tag.value.len == 0) &nil else tag.value,
                        };

                        try self.setNode(leaf);

                        if (utils.lessThan(first_child, tag.key)) {
                            if (leaf.isBoundary()) {
                                try self.new_siblings.append(tag.key);
                            }

                            return Result.update;
                        } else if (utils.equal(first_child, tag.key)) {
                            if (first_child == null or leaf.isBoundary()) {
                                return Result.update;
                            } else {
                                return Result.delete;
                            }
                        } else {
                            return error.InvalidDatabase;
                        }
                    },
                    .delete => |key| {
                        try self.deleteNode(0, key);
                        if (utils.equal(key, first_child)) {
                            return Result.delete;
                        } else {
                            return Result.update;
                        }
                    },
                }
            }

            fn applyNode(self: *Operation, level: u8, first_child: ?[]const u8) !Result {
                try self.log("insertNode({d}, {s})", .{ level, fmtKey(first_child) });

                try self.logger.indent();
                defer self.logger.deindent();

                if (level == 0) {
                    return self.applyLeaf(first_child);
                }

                const key = switch (self.tag) {
                    .set => |entry| entry.key,
                    .delete => |key| key,
                };

                const target = try self.findTargetKey(level, first_child, key);
                try self.log("target: {s}", .{fmtKey(target)});

                const is_left_edge = first_child == null;
                try self.log("is_left_edge: {any}", .{is_left_edge});

                const is_first_child = utils.equal(target, first_child);
                try self.log("is_first_child: {any}", .{is_first_child});

                const result = try self.applyNode(level - 1, target);
                switch (result) {
                    Result.delete => try self.log("result: delete", .{}),
                    Result.update => try self.log("result: update", .{}),
                }

                try self.log("new siblings: {d}", .{self.new_siblings.items.len});
                for (self.new_siblings.items) |child| {
                    try self.log("- {s}", .{fmtKey(child)});
                }

                switch (result) {
                    Result.delete => {
                        assert(!is_left_edge or !is_first_child);

                        // delete the entry and move to the previous child
                        const previous_child_key = try self.moveToPreviousChild(level, target);
                        try self.log("previous_child_key: {s}", .{fmtKey(previous_child_key)});

                        try self.promote(level);

                        const is_previous_child_boundary = try self.hashNode(level, previous_child_key);
                        if (is_first_child or utils.lessThan(previous_child_key, first_child)) {
                            if (is_previous_child_boundary) {
                                try self.new_siblings.append(previous_child_key);
                            }

                            return Result.delete;
                        } else if (utils.equal(previous_child_key, first_child)) {
                            if (is_left_edge or is_previous_child_boundary) {
                                return Result.update;
                            } else {
                                return Result.delete;
                            }
                        } else {
                            if (is_previous_child_boundary) {
                                try self.new_siblings.append(previous_child_key);
                            }

                            return Result.update;
                        }
                    },
                    Result.update => {
                        const is_target_boundary = try self.hashNode(level, target);
                        try self.log("is_target_boundary: {any}", .{is_target_boundary});

                        try self.promote(level);

                        // is_first_child means either target's original value was a boundary, or is_left_edge is true.
                        if (is_first_child) {
                            if (is_target_boundary or is_left_edge) {
                                return Result.update;
                            } else {
                                return Result.delete;
                            }
                        } else {
                            if (is_target_boundary) {
                                try self.new_siblings.append(target);
                            }

                            return Result.update;
                        }
                    },
                }
            }

            fn findTargetKey(self: *Operation, level: u8, first_child: ?[]const u8, key: []const u8) !?[]const u8 {
                assert(level > 0);
                const id = level - 1;
                var target: ?[]const u8 = null;
                if (first_child) |bytes| {
                    target = try self.pool.copy(id, bytes);
                }

                _ = try self.cursor.goToNode(level, first_child);
                while (try self.cursor.goToNext()) |next_child_node| {
                    if (next_child_node.key) |next_child_key| {
                        if (utils.lessThan(key, next_child_key)) {
                            break;
                        } else {
                            target = try self.pool.copy(id, next_child_key);
                        }
                    } else {
                        return error.InvalidDatabase;
                    }
                }

                return target;
            }

            fn moveToPreviousChild(self: *Operation, level: u8, target: ?[]const u8) !?[]const u8 {
                assert(level > 0);
                const id = level - 1;

                // delete the entry and move to the previous child

                _ = try self.cursor.goToNode(level, target);
                try self.cursor.deleteCurrentNode();
                if (self.effects) |effects| effects.delete += 1;

                while (try self.cursor.goToPrevious()) |previous_node| {
                    if (previous_node.key) |key| {
                        const previous_key = try self.pool.copy(id, key);
                        if (try self.getNode(level - 1, previous_key)) |previous_child| {
                            if (previous_child.isBoundary()) {
                                return previous_key;
                            }
                        }

                        try self.cursor.deleteCurrentNode();
                        if (self.effects) |effects| effects.delete += 1;
                    } else {
                        return null;
                    }
                }

                return error.InternalError;
            }

            /// Computes and sets the hash of the given node.
            /// Doesn't assume anything about the current cursor position.
            fn hashNode(self: *Operation, level: u8, key: ?[]const u8) !bool {
                try self.log("hashNode({d}, {s})", .{ level, fmtKey(key) });

                var digest = Blake3.init(.{});

                const first = try self.cursor.goToNode(level - 1, key);
                try self.log("- hashing {s} <- {s}", .{ hex(first.hash), fmtKey(key) });
                digest.update(first.hash);

                while (try self.cursor.goToNext()) |next| {
                    if (next.isBoundary()) {
                        break;
                    } else {
                        try self.log("- hashing {s} <- {s}", .{ hex(next.hash), fmtKey(next.key) });
                        digest.update(next.hash);
                    }
                }

                digest.final(&self.hash_buffer);
                try self.log("--------- {s}", .{hex(&self.hash_buffer)});
                try self.log("setting {s} <- ({d}) {s}", .{ hex(&self.hash_buffer), level, fmtKey(key) });
                const node = Node{ .level = level, .key = key, .hash = &self.hash_buffer, .value = null };

                try self.setNode(node);
                return node.isBoundary();
            }

            fn promote(self: *Operation, level: u8) !void {
                var old_index: usize = 0;
                var new_index: usize = 0;
                const new_sibling_count = self.new_siblings.items.len;
                while (old_index < new_sibling_count) : (old_index += 1) {
                    const key = self.new_siblings.items[old_index];
                    const is_boundary = try self.hashNode(level, key);
                    if (is_boundary) {
                        self.new_siblings.items[new_index] = key;
                        new_index += 1;
                    }
                }

                try self.new_siblings.resize(new_index);
            }

            fn getNode(self: *Operation, level: u8, key: ?[]const u8) !?Node {
                const entry_key = try self.encoder.encodeKey(level, key);
                if (try self.txn.get(entry_key)) |entry_value| {
                    return try Node.parse(entry_key, entry_value);
                } else {
                    return null;
                }
            }

            fn setNode(self: *Operation, node: Node) !void {
                const entry = try self.encoder.encode(node);

                if (self.trace) |trace| try trace.append(node);
                if (self.effects) |effects| {
                    if (try self.txn.get(entry.key)) |_| {
                        effects.update += 1;
                    } else {
                        effects.create += 1;
                    }
                }

                try self.txn.set(entry.key, entry.value);
            }

            fn deleteNode(self: *Operation, level: u8, key: ?[]const u8) !void {
                const entry_key = try self.encoder.encodeKey(level, key);

                if (self.effects) |effects| effects.delete += 1;

                try self.txn.delete(entry_key);
            }

            inline fn log(self: *Operation, comptime format: []const u8, args: anytype) !void {
                try self.logger.print(format, args);
            }
        };

        allocator: std.mem.Allocator,
        is_open: bool = false,

        env: lmdb.Environment,
        txn: lmdb.Transaction,
        cursor: Cursor,

        dbi: ?[*:0]const u8,

        log: ?std.fs.File.Writer = null,
        trace: ?*NodeList,
        effects: ?*Effects,

        encoder: NodeEncoder,
        pool: BufferPool,
        new_siblings: std.ArrayList(?[]const u8),
        hash_buffer: [K]u8 = undefined,

        pub fn open(allocator: std.mem.Allocator, tree: *const Tree, options: Options) !Self {
            var transaction: Self = undefined;
            try transaction.init(allocator, tree, options);
            return transaction;
        }

        pub fn init(self: *Self, allocator: std.mem.Allocator, tree: *const Tree, options: Options) !void {
            if (options.dbi) |dbi| {
                const txn = try lmdb.Transaction.open(tree.env, .{ .read_only = true });
                defer txn.abort();
                if (try txn.get(std.mem.span(dbi)) == null) {
                    return error.DatabaseNotFound;
                }
            }

            const read_only = switch (options.mode) {
                .ReadOnly => true,
                .ReadWrite => false,
            };

            const txn = try lmdb.Transaction.open(tree.env, .{ .read_only = read_only, .dbi = options.dbi });
            errdefer txn.abort();

            try Header.validate(txn);

            self.allocator = allocator;
            self.is_open = true;

            self.env = tree.env;
            self.txn = txn;
            self.cursor = try Cursor.open(allocator, txn);

            if (options.dbi) |ptr| {
                const len = std.mem.indexOfSentinel(u8, 0, ptr);
                const dbi = try allocator.allocSentinel(u8, len, 0);
                std.mem.copy(u8, dbi, std.mem.span(ptr));
                self.dbi = dbi.ptr;
            } else {
                self.dbi = null;
            }

            self.log = options.log;
            self.trace = options.trace;
            self.effects = options.effects;

            self.encoder = NodeEncoder.init(allocator);
            self.pool = BufferPool.init(allocator);
            self.new_siblings = std.ArrayList(?[]const u8).init(allocator);
        }

        pub fn abort(self: *Self) void {
            if (self.is_open) {
                self.is_open = false;
                self.cursor.close();
                self.txn.abort();
            }
        }

        pub fn commit(self: *Self) !void {
            if (self.is_open) {
                self.is_open = false;
                self.cursor.close();
                try self.txn.commit();
            } else {
                return error.TransactionClosed;
            }
        }

        fn deinit(self: *Self) void {
            self.new_siblings.deinit();
            self.pool.deinit();
            self.encoder.deinit();
            if (self.dbi) |ptr| {
                self.allocator.free(ptr);
            }
        }

        pub fn get(self: *Self, key: []const u8) !?[]const u8 {
            const entry_key = try self.cursor.encoder.encodeKey(0, key);
            if (try self.txn.get(entry_key)) |entry_value| {
                const node = try Node.parse(entry_key, entry_value);
                if (node.value) |value| {
                    return value;
                } else {
                    return error.InvalidDatabase;
                }
            } else {
                return null;
            }
        }

        pub fn set(self: *Self, key: []const u8, value: []const u8) !void {
            if (key.len == 0) {
                return error.InvalidKey;
            }

            var operation = try Operation.init(self.allocator, self, Operation.Set(key, value));

            errdefer operation.abort();
            try operation.apply();
            try operation.commit();
        }

        pub fn delete(self: *Self, key: []const u8) !void {
            if (key.len == 0) {
                return error.InvalidKey;
            }

            var operation = try Operation.init(self.allocator, self, Operation.Delete(key));

            errdefer operation.abort();
            try operation.apply();
            try operation.commit();
        }

        pub fn getRoot(self: *Self) !Node {
            return try self.cursor.goToRoot();
        }

        pub fn getNode(self: *Self, level: u8, key: ?[]const u8) !?Node {
            const entry_key = try self.cursor.encoder.encodeKey(level, key);
            if (try self.txn.get(entry_key)) |entry_value| {
                return try Node.parse(entry_key, entry_value);
            } else {
                return null;
            }
        }

        pub fn getUserdata(self: *Self) !?[]const u8 {
            return try self.txn.get(&Header.USERDATA_KEY);
        }

        pub fn setUserdata(self: *Self, userdata: ?[]const u8) !void {
            if (userdata) |value| {
                try self.txn.set(&Header.USERDATA_KEY, value);
            } else {
                try self.txn.delete(&Header.USERDATA_KEY);
            }
        }
    };
}

File: okra-browser/src/transaction.ts
import { IDBPDatabase, IDBPTransaction } from "idb"
import { blake3 } from "@noble/hashes/blake3"

import { toIndex, Key, Node, getNode, isIndex } from "./schema.js"
import {
	equalKeys,
	hashEntry,
	isSplit,
	K,
	leafAnchorHash,
	lessThan,
} from "./utils.js"

// we have enums at home
const Result = { Update: 0, Delete: 1 } as const
type Result = typeof Result[keyof typeof Result]

type Operation =
	| { type: "set"; key: Uint8Array; value: Uint8Array }
	| { type: "delete"; key: Uint8Array }

export class ReadOnlyTransaction {
	public open = true

	constructor(
		public readonly db: IDBPDatabase,
		public readonly storeName: string
	) {}

	public close() {
		this.open = false
	}

	public async getRoot(
		options: {
			txn?:
				| IDBPTransaction<unknown, [string], "readwrite">
				| IDBPTransaction<unknown, [string], "readonly">
		} = {}
	): Promise<Node> {
		if (!this.open) {
			throw new Error("transaction closed")
		}

		const txn = options.txn ?? this.db.transaction(this.storeName, "readonly")
		const cursor = await txn.store.openCursor(null, "prev")
		if (cursor === null || !isIndex(cursor.key)) {
			console.error(cursor?.key)
			throw new Error("invalid database")
		}

		const [level, key] = cursor.key
		if (key !== undefined) {
			console.error(key)
			throw new Error("invalid database")
		}

		const { hash, value } = cursor.value
		if (value !== undefined) {
			console.error(value)
			throw new Error("invalid database")
		}

		return { level, key: null, hash }
	}

	public async getNode(
		level: number,
		key: Key,
		options: {
			txn?:
				| IDBPTransaction<unknown, [string], "readwrite">
				| IDBPTransaction<unknown, [string], "readonly">
		} = {}
	): Promise<Node> {
		if (!this.open) {
			throw new Error("transaction closed")
		}

		const txn = options.txn ?? this.db.transaction(this.storeName, "readonly")
		const node = await txn.store.get(toIndex(level, key))
		if (node === undefined) {
			throw new Error("key not found")
		}

		return { level, key, ...node }
	}

	public async get(
		key: Uint8Array,
		options: {
			txn?:
				| IDBPTransaction<unknown, [string], "readwrite">
				| IDBPTransaction<unknown, [string], "readonly">
		} = {}
	): Promise<Uint8Array | null> {
		if (!this.open) {
			throw new Error("transaction closed")
		}

		const txn = options.txn ?? this.db.transaction(this.storeName, "readonly")
		const node = await txn.store.get([0, key])
		if (node === undefined) {
			return null
		} else if (node.value === undefined) {
			throw new Error("invalid database")
		} else {
			return node.value
		}
	}

	public async seek(
		level: number,
		key: Key,
		options: {
			txn?:
				| IDBPTransaction<unknown, [string], "readwrite">
				| IDBPTransaction<unknown, [string], "readonly">
		} = {}
	): Promise<Node | null> {
		if (!this.open) {
			throw new Error("transaction closed")
		}

		const txn = options.txn ?? this.db.transaction(this.storeName, "readonly")
		const range = IDBKeyRange.lowerBound(toIndex(level, key))
		let cursor = await txn.store.openCursor(range, "next")
		if (cursor === null) {
			return null
		}

		const node = getNode(cursor)
		return node.level === level ? node : null
	}

	public async getChildren(
		level: number,
		key: Key,
		options: {
			txn?:
				| IDBPTransaction<unknown, [string], "readwrite">
				| IDBPTransaction<unknown, [string], "readonly">
		} = {}
	): Promise<Node[]> {
		if (!this.open) {
			throw new Error("transaction closed")
		}

		if (level === 0) {
			throw new Error("cannot get children of a leaf node")
		}

		const children: Node[] = []
		for await (const node of this.range(level - 1, key, options)) {
			children.push(node)
		}

		return children
	}

	public async *range(
		level: number,
		firstChild: Key,
		options: {
			txn?:
				| IDBPTransaction<unknown, [string], "readwrite">
				| IDBPTransaction<unknown, [string], "readonly">
		} = {}
	): AsyncIterable<Node> {
		if (!this.open) {
			throw new Error("transaction closed")
		}

		const txn = options.txn ?? this.db.transaction(this.storeName, "readonly")
		const range = IDBKeyRange.lowerBound(toIndex(level, firstChild))
		let cursor = await txn.store.openCursor(range, "next")
		if (cursor === null) {
			throw new Error("invalid database")
		}

		let node = getNode(cursor)
		if (node.level !== level || !equalKeys(node.key, firstChild)) {
			console.error("looking for", level, firstChild, "and got", node)
			throw new Error("key not found")
		}

		yield node

		while ((cursor = await cursor.continue())) {
			const node = getNode(cursor)
			if (node.level !== level || isSplit(node.hash)) {
				break
			} else {
				yield node
			}
		}
	}
}

export class ReadWriteTransaction extends ReadOnlyTransaction {
	private newSiblings: Key[] = []

	public async reset() {
		if (!this.open) {
			throw new Error("transaction closed")
		}

		await this.db.clear(this.storeName)
		await this.db.put(this.storeName, { hash: leafAnchorHash }, [0])
	}

	public async set(key: Uint8Array, value: Uint8Array): Promise<void> {
		if (!this.open) {
			throw new Error("transaction closed")
		}

		const txn = this.db.transaction(this.storeName, "readwrite")
		try {
			await this.apply(txn, { type: "set", key, value })
		} catch (err) {
			txn.abort()
			throw err
		}

		txn.commit()
	}

	public async delete(key: Uint8Array): Promise<void> {
		if (!this.open) {
			throw new Error("transaction closed")
		}

		const txn = this.db.transaction(this.storeName, "readwrite")
		try {
			await this.apply(txn, { type: "delete", key })
		} catch (err) {
			txn.abort()
			throw err
		}

		txn.commit()
	}

	private async apply(
		txn: IDBPTransaction<unknown, [string], "readwrite">,
		operation: Operation
	): Promise<void> {
		if (this.newSiblings.length !== 0) {
			throw new Error("internal error")
		}

		const root = await this.getRoot({ txn })

		const result =
			root.level === 0
				? await this.applyLeaf(txn, null, operation)
				: await this.applyNode(txn, root.level - 1, null, operation)

		if (result === Result.Delete) {
			throw new Error("internal error")
		}

		let rootLevel = root.level || 1

		await this.hashNode(txn, rootLevel, null)

		while (this.newSiblings.length > 0) {
			await this.promote(txn, rootLevel)
			rootLevel += 1
			await this.hashNode(txn, rootLevel, null)
		}

		while (rootLevel > 0) {
			const { key: last } = await this.getLast(txn, rootLevel - 1)
			if (last !== null) {
				break
			} else {
				await txn.store.delete(toIndex(rootLevel, null))
			}

			rootLevel--
		}
	}

	private async applyLeaf(
		txn: IDBPTransaction<unknown, [string], "readwrite">,
		firstChild: Key,
		operation: Operation
	): Promise<Result> {
		if (operation.type === "set") {
			const leaf: Node = {
				level: 0,
				key: operation.key,
				hash: hashEntry(operation.key, operation.value),
				value: operation.value,
			}

			await this.setNode(txn, leaf)
			if (lessThan(firstChild, operation.key)) {
				if (isSplit(leaf.hash)) {
					this.newSiblings.push(operation.key)
				}

				return Result.Update
			} else if (equalKeys(firstChild, operation.key)) {
				if (firstChild === null || isSplit(leaf.hash)) {
					return Result.Update
				} else {
					return Result.Delete
				}
			} else {
				throw new Error("invalid database")
			}
		} else if (operation.type === "delete") {
			await txn.store.delete(toIndex(0, operation.key))
			if (equalKeys(operation.key, firstChild)) {
				return Result.Delete
			} else {
				return Result.Update
			}
		} else {
			throw new Error("invalid operation type")
		}
	}

	private async applyNode(
		txn: IDBPTransaction<unknown, [string], "readwrite">,
		level: number,
		firstChild: Key,
		operation: Operation
	): Promise<Result> {
		if (level === 0) {
			return this.applyLeaf(txn, firstChild, operation)
		}

		const target = await this.findTarget(txn, level, firstChild, operation.key)

		const isLeftEdge = firstChild === null
		const isFirstChild = equalKeys(target, firstChild)

		const result = await this.applyNode(txn, level - 1, target, operation)
		if (result === Result.Delete) {
			if (isLeftEdge && isFirstChild) {
				throw new Error("internal error")
			}

			const previousChild = await this.moveToPreviousChild(txn, level, target)

			await this.promote(txn, level)

			const isPreviousChildSplit = await this.hashNode(
				txn,
				level,
				previousChild
			)

			if (isFirstChild || lessThan(previousChild, firstChild)) {
				if (isPreviousChildSplit) {
					this.newSiblings.push(previousChild)
				}

				return Result.Delete
			} else if (equalKeys(previousChild, firstChild)) {
				if (isLeftEdge || isPreviousChildSplit) {
					return Result.Update
				} else {
					return Result.Delete
				}
			} else {
				if (isPreviousChildSplit) {
					this.newSiblings.push(previousChild)
				}

				return Result.Update
			}
		} else {
			const isTargetSplit = await this.hashNode(txn, level, target)

			await this.promote(txn, level)

			if (isFirstChild) {
				if (isTargetSplit || isLeftEdge) {
					return Result.Update
				} else {
					return Result.Delete
				}
			} else {
				if (isTargetSplit) {
					this.newSiblings.push(target)
				}

				return Result.Update
			}
		}
	}

	private async promote(
		txn: IDBPTransaction<unknown, [string], "readwrite">,
		level: number
	): Promise<void> {
		const newSiblings: Key[] = []
		for (const newChild of this.newSiblings) {
			const isSplit = await this.hashNode(txn, level, newChild)
			if (isSplit) {
				newSiblings.push(newChild)
			}
		}

		this.newSiblings = newSiblings
	}

	private async findTarget(
		txn: IDBPTransaction<unknown, [string], "readwrite">,
		level: number,
		firstChild: Key,
		key: Uint8Array
	): Promise<Key> {
		let target: Node | undefined = undefined
		for await (const node of this.range(level, firstChild, { txn })) {
			if (lessThan(key, node.key)) {
				break
			} else {
				target = node
			}
		}

		if (target === undefined) {
			throw new Error("invalid database")
		} else {
			return target.key
		}
	}

	// deletes the target key and moves backwards until it finds a new one.
	// this method is responsible for some magic so be really careful.
	private async moveToPreviousChild(
		txn: IDBPTransaction<unknown, [string], "readwrite">,
		level: number,
		target: Key
	): Promise<Key> {
		if (level === 0) {
			throw new Error("internal error")
		}

		const index = toIndex(level, target)
		await txn.store.delete(index)
		const range = IDBKeyRange.upperBound(index, true)
		let cursor = await txn.store.openCursor(range, "prev")
		while (cursor !== null) {
			const { key: previousChild } = getNode(cursor)
			if (previousChild === null) {
				return null
			}

			const previousGrandChild = await this.getNode(level - 1, previousChild, {
				txn,
			})
			if (isSplit(previousGrandChild.hash)) {
				return previousChild
			}

			await txn.store.delete(toIndex(level, previousChild))
			cursor = await cursor.continue()
		}

		throw new Error("internal error")
	}

	// Computes and sets the hash of the given node.
	// Doesn't assume anything about the current cursor position.
	// Returns isSplit for the updated hash.
	private async hashNode(
		txn: IDBPTransaction<unknown, [string], "readwrite">,
		level: number,
		key: Key
	): Promise<boolean> {
		const hash = blake3.create({ dkLen: K })
		for await (const node of this.range(level - 1, key, { txn })) {
			hash.update(node.hash)
		}

		const node: Node = { level, key, hash: hash.digest() }
		await this.setNode(txn, node)
		return isSplit(node.hash)
	}

	private async setNode(
		txn: IDBPTransaction<unknown, [string], "readwrite">,
		{ level, key, ...node }: Node
	): Promise<void> {
		await txn.store.put(node, toIndex(level, key))
	}

	private async getLast(
		txn: IDBPTransaction<unknown, [string], "readwrite">,
		level: number
	): Promise<Node> {
		const range = IDBKeyRange.upperBound(toIndex(level + 1, null), true)
		const cursor = await txn.store.openCursor(range, "prev")
		if (cursor === null) {
			throw new Error("invalid database")
		} else {
			return getNode(cursor)
		}
	}
}
